Adding rotationAlignment: 'horizon' for markers on globe

SnailBones opened the pull request:
Takeaway from the discussion in https://github.com/mapbox/mapbox-gl-js/pull/11642 is that a globe-oriented rotation alignment property is useful (but not a pitch alignment), and the clearest name for the property is "horizon." This PR adds that feature accordingly.

```js
const marker = new mapboxgl.Marker({
      rotationAlignment: 'horizon',
      pitchAlignment: 'viewport' // This line can also be left out.
  })

```
![image](https://user-images.githubusercontent.com/14878684/168395033-ff5af516-238c-498e-a92e-9c81c5717cd4.png)


As the map zooms in and transitions from globe to Mercator, markers smoothly transition into `rotationAlignment: viewport` behavior.

https://user-images.githubusercontent.com/14878684/168399083-85d0f22c-2e74-4f55-8bcb-b6ef8ea4f657.mov



With custom pin markers:

https://user-images.githubusercontent.com/14878684/168398499-424e6f57-2831-4e86-9383-58fb8c831968.mov




## Launch Checklist

<!-- Thanks for the PR! Feel free to add or remove items from the checklist. -->

 - [X] briefly describe the changes in this PR
 - [x] include before/after visuals or gifs if this PR includes visual changes
 - [x] write tests for all new functionality
 - [x] document any changes to public APIs
 - [x] manually test the debug page
 - [X] tagged `@mapbox/map-design-team` `@mapbox/static-apis` if this PR includes style spec API or visual changes
 - [X] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog>Adding new marker styling option: rotationAlignment: 'horizon' allowing marker rotation to match the curvature of the horizon in globe view</changelog>`
 
 cc @mapbox/map-design-team

karimnaaji commented on src/ui/marker.js:
```diff
@@ -790,6 +817,7 @@ export default class Marker extends Evented {
      */
     getPitchAlignment(): string {
-        return this._pitchAlignment === `auto` ? this.getRotationAlignment() : this._pitchAlignment;
+        if (this._map && this._pitchAlignment === 'horizon' && !this._map._usingGlobe()) { return 'viewport'; }
+        return this._pitchAlignment === 'auto' ? this.getRotationAlignment() : this._pitchAlignment;
```
These checks are starting to become difficult to follow and ready. Partly because `_usingGlobe` is a misleading name. To make it self-documenting, I think we should move away from referring that's it's about _using_ globe and clarify that it's whether the map is actively _showing_, so how about:
1. Renaming `map._usingGlobe` to `map._showingGlobe()`
2. Adding a helper function:
```
_usingHorizonAlignment(): boolean {
    return this._pitchAlignment === 'horizon' && this._map && !this._map._showingGlobe();
}
```
3. Expanding these checks for clarity:
```suggestion
        if (this._usingHorizonAlignment()) {
            return 'viewport';
        }
        if (this._pitchAlignment === 'auto') {
            return this.getRotationAlignment();
        }
        return this._pitchAlignment;
```

SnailBones replied:
Good call Karim! I agree that `map._showingGlobe()` is much clearer.

karimnaaji commented on src/ui/marker.js:
```diff
@@ -765,7 +790,9 @@ export default class Marker extends Evented {
      */
     getRotationAlignment(): string {
-        return this._rotationAlignment === `auto` ? 'viewport' : this._rotationAlignment;
+        return (this._rotationAlignment === `auto` ||
+            (this._map && this._rotationAlignment === 'horizon' && !this._map._usingGlobe())) ?
+            'viewport' : this._rotationAlignment;
```
Same suggestion as the other comment, this tries to do a bit much, to add some clarity we can expand as follow:

```suggestion
        if (this._rotationAlignment === `auto` || this._usingHorizonAlignment()) {
            return 'viewport';
        }
        return this._rotationAlignment;
```

karimnaaji commented on src/ui/marker.js:
```diff
@@ -500,20 +511,34 @@ export default class Marker extends Evented {
             }
-            return this._rotation - this._map.getBearing();
+            return this._rotation - map.getBearing();
+        } else if (this._rotationAlignment === "horizon" && map._usingGlobe()) {
```
Similar suggestion as the other comment:
```suggestion
        } else if (this._usingHorizonAlignment()) {
```

karimnaaji commented on src/ui/marker.js:
```diff
@@ -479,19 +487,22 @@ export default class Marker extends Evented {
-        const xTilt = -posFromCenter.y * tiltOverDist;
-        if (!xTilt && !yTilt) { return ''; }
-        return `rotateX(${xTilt}deg) rotateY(${yTilt}deg)`;
+        return ''; // Invalid alignments behave as viewport
```
A few comments to improve clarity for that function:
1. Do error checks first instead of last, the validation of inputs is the most important thing when reading what a function does. e.g. 
```
function() {
    if (bad_arguments || error) {
        return '';
    }
    ...
}
```
2. For checks that aren't trivial (e.g. one liner), try to not return in the middle of the function to consolidate the control flow to only a few return statements, so that it's easier to follow what the function does. Otherwise it makes it really hard to read and can be a source of bugs when new statements are added.
```
        if (map && pos && alignment === 'map') {
            if (!map._usingGlobe()) { // 'map' alignment on a flat map
                const pitch = map.getPitch();
                return pitch ? `rotateX(${pitch}deg)` : '';
            } // "map" alignment on globe
            const tilt = radToDeg(globeTiltAtLngLat(map.transform, this._lngLat));
            const posFromCenter = pos.sub(globeCenterToScreenPoint(map.transform));
            const tiltOverDist =  tilt / (Math.abs(posFromCenter.x) + Math.abs(posFromCenter.y));
            const yTilt = posFromCenter.x * tiltOverDist;
            const xTilt = -posFromCenter.y * tiltOverDist;
            if (!xTilt && !yTilt) { return ''; }
            return `rotateX(${xTilt}deg) rotateY(${yTilt}deg)`;
        }
        return ''; // Invalid alignments behave as viewport
```
This block is too dense and has 4 different return statements, we can make the condition more explicit, do error check first and store the result in a variable if needed, something like:
```
function() {
    let result = default;
    if (error) {
        return result;
    }
    
    if (...) {
        result = ...
    } else if (...) {
        result = ...
    }
    return result;
}
```

3. Prefer multi line comment instead of same line comments, e.g.:
```
if (!map._usingGlobe()) {
    // 'map' alignment on a flat map
    ...
}
// "map" alignment on globe
```
instead of
```
if (!map._usingGlobe()) { // 'map' alignment on a flat map
    ...
}  // "map" alignment on globe
```

karimnaaji commented on src/geo/projection/globe_util.js:
```diff
@@ -339,6 +340,7 @@ function csLatLngToECEF(cosLat: number, sinLat: number, lng: number, radius: num
 }
 
 export function latLngToECEF(lat: number, lng: number, radius?: number): Array<number> {
+    assert(lat <= 90 && lat >= -90, 'Lattitude must be between -90 and 90');
```
👍

karimnaaji reviewed (commented):
Nice addition! This will be a good feature to land for 2.9.1 🎉

I have a few stylistic comments regarding the newly added code. Otherwise, I haven't been able to locally test `markers-custom`, it seems like the page doesn't show the markers (only popups):

<img width="813" alt="Screen Shot 2022-05-16 at 12 53 31 PM" src="https://user-images.githubusercontent.com/7061573/168671558-158c5750-0952-4a90-bd20-6e947f36fa95.png">

But it worked great and the other page `markers`

karimnaaji commented on src/ui/marker.js:
```diff
@@ -479,19 +487,22 @@ export default class Marker extends Evented {
+            const tiltOverDist =  tilt / (Math.abs(posFromCenter.x) + Math.abs(posFromCenter.y));
+            const yTilt = posFromCenter.x * tiltOverDist;
+            const xTilt = -posFromCenter.y * tiltOverDist;
+            if (!xTilt && !yTilt) { return ''; }
```
`!xTilt && !yTilt` are these condition checking against whether the variable is defined, or whether it's the value `0`? I assume it's against `0`? If so, we should prefer being being explicit

SnailBones replied:
Turns out this was both checking against 0 and `undefined` values created by the division by 0 as you mention in your previous comment. Updating the code to be more clear.

karimnaaji commented on src/ui/marker.js:
```diff
@@ -479,19 +487,22 @@ export default class Marker extends Evented {
+            } // "map" alignment on globe
+            const tilt = radToDeg(globeTiltAtLngLat(map.transform, this._lngLat));
+            const posFromCenter = pos.sub(globeCenterToScreenPoint(map.transform));
+            const tiltOverDist =  tilt / (Math.abs(posFromCenter.x) + Math.abs(posFromCenter.y));
```
Should we check against a possible division by `0` to be safe?

SnailBones commented:
> I have a few stylistic comments regarding the newly added code.

@karimnaaji I realized there was still a bit of vestigial logic accounting for the experimental horizon *pitch* alignment in #1642, so I've been able to simplify the code in a few places. It now seems that adding a`this._enableHorizonAlignment()` function wouldn't need to be used more than once, otherwise I've addressed your concerns.

> Otherwise, I haven't been able to locally test markers-custom, it seems like the page doesn't show the markers (only popups):

Whoops, forgot to commit the custom icon! 🤦 That's now fixed.

karimnaaji commented on src/ui/map.js:
```diff
@@ -1139,11 +1139,11 @@ class Map extends Camera {
      * }
      */
-    _usingGlobe(): boolean { return this.transform.projection.name === 'globe'; }
+    _showingGlobe(): boolean { return this.transform.projection.name === 'globe'; }
```
nit: Might be worth adding a small unit test to this function.

karimnaaji commented on src/ui/marker.js:
```diff
@@ -28,7 +28,13 @@ type Options = {
+
 export const TERRAIN_OCCLUDED_OPACITY = 0.2;
+// Zoom levels to transition 'horizon' aligned markers in globe view.
+const ALIGN_TO_HORIZON_BELOW_ZOOM = 4;
+const ALIGN_TO_SCREEN_ABOVE_ZOOM = 6; // Can't be larger than GLOBE_ZOOM_THRESHOLD_MAX.
```
Why not use the transition values from globe? Or at least making it a function of it (e.g. `ZOOM_TRANSITION_THRESHOLD - 1` or else...), so that changing the setting of the transition happens in the same place. Otherwise this is a bit fragile to change when updating the actual globe-mercator transition zoom threshold.

SnailBones replied:
> Why not use the transition values from globe? Or at least making it a function of it

That's not a bad idea!I think that if we increase the globe transition we don't necessarily want to change these values, they're relatively arbitrary. It's mostly coincidental that I found that IMO they look good matching the current globe transition.

I think I like your idea of an assertion!

karimnaaji commented on src/ui/marker.js:
```diff
@@ -120,8 +126,6 @@ export default class Marker extends Evented {
             this._element = DOM.create('div');
 
             // create default map marker SVG
-            const defaultHeight = 41;
-            const defaultWidth = 27;
```
nit: Why move these out? I'd prefer locality, so the values aren't few hundreds of code lines away.

SnailBones replied:
My thought was that this makes them less magic-numbery and more consistent with the other constants defined at the start of the file, but that's a good point. Do you think that `ALIGN_TO_HORIZON_BELOW_ZOOM ` and `ALIGN_TO_HORIZON_ABOVE_ZOOM` should also be defined closer to where they are used?

karimnaaji replied:
>Do you think that ALIGN_TO_HORIZON_BELOW_ZOOM and ALIGN_TO_HORIZON_ABOVE_ZOOM should also be defined closer to where they are used?

I think so, since they're not really shared by anything else and not exported for use outside of this file, it's probably good to keep it local to the block they're used in as well instead of in the global scope. Being consistent in naming would also make sense if you keep here, e.g. making these two variables all caps similar to the other ones.

karimnaaji commented on src/ui/marker.js:
```diff
@@ -500,20 +519,34 @@ export default class Marker extends Evented {
             }
-            return this._rotation - this._map.getBearing();
+            return this._rotation - map.getBearing();
+        } else if (this._rotationAlignment === "horizon" && map._showingGlobe()) {
```
Why is this not checking against `alignment`? e.g. `alignment === "horizon"`, isn't `getRotationAlignment` already checking whether the map is showing globe so this test becomes redundant?

SnailBones replied:
Good catch!

karimnaaji commented on src/ui/marker.js:
```diff
@@ -500,20 +519,34 @@ export default class Marker extends Evented {
-            if (pos && map && map._usingGlobe()) {
+        const pos = this._pos;
+        const map = this._map;
+        if (!map || !pos) { return 0; }
+        const alignment = this.getRotationAlignment();
+        if (alignment === 'map') {
+            if (map._showingGlobe()) {
                 const north = map.project(new LngLat(this._lngLat.lng, this._lngLat.lat + .001));
                 const south = map.project(new LngLat(this._lngLat.lng, this._lngLat.lat - .001));
                 const diff = south.sub(north);
                 return this._rotation + radToDeg(Math.atan2(diff.y, diff.x)) - 90;
             }
-            return this._rotation - this._map.getBearing();
+            return this._rotation - map.getBearing();
+        } else if (this._rotationAlignment === "horizon" && map._showingGlobe()) {
+            const zoom = map.getZoom();
+            const centerPoint = globeCenterToScreenPoint(map.transform);
+            let zoomTransition = 1;
+            if (zoom > ALIGN_TO_HORIZON_BELOW_ZOOM) {
+                const smooth = smoothstep(ALIGN_TO_HORIZON_BELOW_ZOOM, ALIGN_TO_SCREEN_ABOVE_ZOOM, zoom);
+                centerPoint.y += smooth * map.transform.height;
+                zoomTransition = 1 - smooth;
+            }
+            const rel = pos.sub(centerPoint);
+            const angle = radToDeg(Math.atan2(rel.y, rel.x));
+            const up = angle > 90 ? angle - 270 : angle + 90;
+            return up * zoomTransition + this._rotation;
+
         }
-        return 0;
+        return this._rotation;
```
I think this can be simplified a bit (untested so please only take that as a starting point!):

- Reduce number of return
- Store the result in an offset variable, since the result can be folded as `rotation + offset`, also make the logic easier to reason about (the default behavior is _no offset_, having an anchor _applies an offset_ to the initial rotation)
- Remove an extra `if`, since the `smoothstep` is already returning 0 for values below `ALIGN_TO_HORIZON_BELOW_ZOOM` and its evaluation is trivial enough that it's not worth adding an extra branch
```js
let offset = 0;
if (alignment === 'map') {
    if (map._showingGlobe()) {
        const north = map.project(new LngLat(this._lngLat.lng, this._lngLat.lat + .001));
        const south = map.project(new LngLat(this._lngLat.lng, this._lngLat.lat - .001));
        const diff = south.sub(north);
        offset = radToDeg(Math.atan2(diff.y, diff.x)) - 90;
    } else {
        offset = -map.getBearing();
    }
} else if (this._rotationAlignment === "horizon" && map._showingGlobe()) {
    const centerPoint = globeCenterToScreenPoint(map.transform);
    const smooth = smoothstep(ALIGN_TO_HORIZON_BELOW_ZOOM, ALIGN_TO_SCREEN_ABOVE_ZOOM, map.getZoom());
    centerPoint.y += smooth * map.transform.height;
    const zoomTransition = 1 - smooth;
    const rel = pos.sub(centerPoint);
    const angle = radToDeg(Math.atan2(rel.y, rel.x));
    const up = angle > 90 ? angle - 270 : angle + 90;
    offset = up * zoomTransition;
}
return this._rotation + offset;
```

SnailBones replied:
Good ideas Karim! I've implemented those changes and also combined the function into `_calculateZTransform()`
